### PR TITLE
fix: use shell AST for DIP125 binary detection (tracker#87)

### LINKED ISSUE #87: Bash variables treated as executables
Having node like this:

```
  tool IncrementCounter
    label: "Increment Iteration"
    timeout: 5s
    command:
      set -eu
      COUNTER='.ai/ralph/iteration-count.txt'
      count=0
      [ -f "$COUNTER" ] && count=$(cat "$COUNTER")
      count=$((count + 1))
      printf '%s' "$count" > "$COUNTER"
      printf '%s' "$count"
```

produces warnings like these:

```
hint[DIP125]: tool command binary "COUNTER='.ai/ralph/iteration-count.txt'" not found on PATH
  --> specs/001-core/pipeline.dip:64:24
hint[DIP125]: tool command binary "count=$(cat" not found on PATH
  --> specs/001-core/pipeline.dip:122:19
```

My expectation is to not have warnings in these cases.

### PR DESCRIPTION
**Pull request by clintecker**
## Summary

Replaces the regex/first-token approach for DIP125 (tool command binary not found) with proper shell AST parsing via `mvdan.cc/sh/v3`.

**Before:** `COUNTER='.ai/count.txt'` and `count=$(cat file)` were reported as missing binaries because the linter naively took the first whitespace-delimited token of each line.

**After:** The shell command is parsed into an AST. `CallExpr` nodes with `Args` are real command invocations; pure assignments and command substitutions are correctly identified.

Fixes 2389-research/tracker#87

## Test plan
- [x] `just check` passes
- [x] Regression tests for variable assignments, command substitution, pipes, conditionals
- [x] Integration test with the exact command from tracker#87 — zero false positives
- [x] `just lint-examples` — only legitimate hints remain (swift, command)

## What changed
- Added `mvdan.cc/sh/v3` dependency (pure Go shell parser, powers `shfmt`)
- Replaced `extractBinary`/`firstToken`/`shellPreamble` regex with `syntax.Walk` AST traversal
- Added `set`, `cd`, `export`, `unset` to shell builtins list

## Summary by CodeRabbit

* **Bug Fixes & Improvements**
  * Improved shell command parsing to more accurately extract/validate binaries from complex shell constructs.

* **New Features**
  * Added automated spec verification and a dedicated release-check routine to the task runner.

* **Tests**
  * Broadened and reorganized command-parsing tests to cover more shell patterns and edge cases.

* **Documentation**
  * Spec updated to document a new optional workflow-level vars block.

* **Chores**
  * Updated dependency versions and tightened pre-commit checks (short-mode tests and generated-artifact verification).

### CODE REVIEW

File: validator/lint_tool_helpers.go [hunk 1]
```diff
@@ -42,30 +43,57 @@ func checkToolCtxVars(n *ir.Node) []Diagnostic {
...
+// command binary into bin.
+func walkForBinary(bin *string) func(syntax.Node) bool {
+	return func(node syntax.Node) bool {
+		if *bin != "" {
+			return false
+		}
+		name := callExprBinary(node)
+		if name != "" && !isShellBuiltin(name) {
+			*bin = name
+			return false
```
[the remark this answers is not in the document]
Comment (clintecker) [reply validator/lint_tool_helpers.go]: Fixed in dadf859. Added `preambleCommands` set (mkdir, touch, chmod, rm, cp, mv) and `isSkippableCommand` that checks both builtins and preamble. The walker skips these and continues to find the real tool binary.

File: docs/superpowers/plans/2026-04-16-vars-block.md [hunk 2]
```diff
@@ -0,0 +1,812 @@
...
+
+**Architecture:** `Workflow.Vars map[string]string` is a new top-level IR field. The parser handles `vars` as a workflow-level block (peer to `defaults`, `edges`). The formatter emits it between defaults and nodes. DOT export emits vars as graph-level attributes. DOT migrate captures unknown graph attributes into Vars.
+
+**Tech Stack:** Go, dippin-lang IR/parser/formatter/export/migrate packages
+
+---
+
+## File Map
+
+| File | Action | Responsibility |
```
[the remark this answers is not in the document]
Comment (clintecker) [reply docs/superpowers/plans/2026-04-16-vars-block.md:15]: The plan doc is committed to main before this branch point — it's not part of this PR's changes, just visible in the diff because the base has diverged. It ships in the vars PR (#15).

File: validator/lint_tool_helpers.go [hunk 3]
```diff
@@ -77,6 +105,7 @@ var shellBuiltins = map[string]bool{
 	"exec": true, "exit": true, "return": true, "shift": true,
 	"trap": true, "wait": true, "true": true, "false": true,
 	"source": true, ".": true, "local": true, "declare": true,
+	"set": true, "cd": true, "export": true, "unset": true,
 }
```
[the remark this answers is not in the document]
Comment (clintecker) [reply validator/lint_tool_helpers.go:158]: Fixed in dadf859. Added `command` to `shellBuiltins`.

File: validator/lint_tool_helpers.go [hunk 4]
```diff
@@ -42,30 +43,57 @@ func checkToolCtxVars(n *ir.Node) []Diagnostic {
...
+		if *bin != "" {
+			return false
+		}
+		name := callExprBinary(node)
+		if name != "" && !isShellBuiltin(name) {
+			*bin = name
+			return false
 		}
-		return firstToken(line)
+		return true
```
[the remark this answers is not in the document]
Comment (clintecker) [reply validator/lint_tool_helpers.go]: Fixed in dadf859. `mkdir` is in `preambleCommands` — the walker skips it and finds the next substantive binary.

File: validator/lint_tool_cmd_test.go [hunk 5]
```diff
@@ -168,17 +168,48 @@ func TestExtractBinary(t *testing.T) {
...
+		{"FOO=bar", ""},
+		{"FOO=bar\nBAZ=qux", ""},
+		// Assignment with command on same line
+		{"FOO=bar ls -la", "ls"},
+		// Command substitution in assignment — walk finds cat inside $() first
+		{"count=$(cat file)\necho $count", "cat"},
+		// Pipes
+		{"cat file | grep pattern", "cat"},
+		// Conditional
+		{"if true; then echo yes; fi", ""},
```
[the remark this answers is not in the document]
Comment (clintecker) [reply validator/lint_tool_cmd_test.go]: Fixed in dadf859. Added test cases for `mkdir -p` preamble, `command -v`, and combined patterns.

File: validator/lint_tool_helpers.go [hunk 6]
```diff
@@ -42,30 +43,57 @@ func checkToolCtxVars(n *ir.Node) []Diagnostic {
...
 
-// shellPreamble patterns that should be skipped when finding the binary.
-var shellPreamble = regexp.MustCompile(
-	`^(set\s+-\w+|cd\s+\S+|export\s+\S+|mkdir\s+-p\s+\S+|#.*)$`,
-)
-
-// extractBinary finds the first non-preamble command's binary name.
+// extractBinary parses a shell command and returns the first non-builtin
+// command name. Uses a proper shell AST parser to correctly handle
+// variable assignments, pipes, subshells, command substitution, etc.
```
[the remark this answers is not in the document]
Comment (clintecker) [reply validator/lint_tool_helpers.go]: Fixed in 4d5bed3. Doc comment now mentions preamble skipping.

File: validator/lint_tool_cmd_test.go [hunk 7]
```diff
@@ -168,17 +168,54 @@ func TestExtractBinary(t *testing.T) {
...
+		// Preamble commands skipped
+		{"mkdir -p .ai/cache\nshellcheck script.sh", "shellcheck"},
+		{"mkdir -p /tmp/out\ntouch /tmp/out/file\ncurl http://x", "curl"},
+		// command builtin
+		{"if command -v shellcheck >/dev/null 2>&1; then shellcheck script.sh; fi", "shellcheck"},
+		{"command -v git && git status", "git"},
```
[the remark this answers is not in the document]
Comment (clintecker) [reply validator/lint_tool_cmd_test.go]: Fixed in 4d5bed3. Added heredoc and arithmetic test cases. The parser handles these correctly — heredoc returns `cat`, arithmetic returns empty.

File: validator/lint_tool_cmd_test.go [hunk 8]
```diff
@@ -168,17 +168,54 @@ func TestExtractBinary(t *testing.T) {
...
-		got := extractBinary(tt.cmd)
-		if got != tt.want {
-			t.Errorf("extractBinary(%q) = %q, want %q", tt.cmd, got, tt.want)
-		}
+		t.Run(tt.cmd, func(t *testing.T) {
+			got := extractBinary(tt.cmd)
+			if got != tt.want {
+				t.Errorf("extractBinary(%q) = %q, want %q", tt.cmd, got, tt.want)
+			}
+		})
```
[the remark this answers is not in the document]
Comment (clintecker) [reply validator/lint_tool_cmd_test.go:201]: Fixed in 4d5bed3. Subtests now use descriptive names like `var_assign_then_echo` instead of raw shell commands.

File: docs/superpowers/plans/2026-04-16-vars-block.md [hunk 9]
```diff
@@ -0,0 +1,812 @@
+# Vars Block Implementation Plan
+
+> **For agentic workers:** REQUIRED SUB-SKILL: Use superpowers:subagent-driven-development (recommended) or superpowers:executing-plans to implement this plan task-by-task. Steps use checkbox (`- [ ]`) syntax for tracking.
+
+**Goal:** Add a `vars` block to Dippin workflows for declaring user-defined key-value variables that export as DOT graph attributes.
+
+**Architecture:** `Workflow.Vars map[string]string` is a new top-level IR field. The parser handles `vars` as a workflow-level block (peer to `defaults`, `edges`). The formatter emits it between defaults and nodes. DOT export emits vars as graph-level attributes. DOT migrate captures unknown graph attributes into Vars.
```
[the remark this answers is not in the document]
Comment (clintecker) [reply docs/superpowers/plans/2026-04-16-vars-block.md:7]: The plan doc is on main before this branch point — not part of this PR's changes.

File: validator/lint_tool_helpers.go [hunk 10]
```diff
@@ -42,30 +43,59 @@ func checkToolCtxVars(n *ir.Node) []Diagnostic {
...
+// walkForBinary returns a walk function that captures the first non-builtin
+// command binary into bin.
+func walkForBinary(bin *string) func(syntax.Node) bool {
+	return func(node syntax.Node) bool {
+		if *bin != "" {
+			return false
+		}
+		name := callExprBinary(node)
+		if name != "" && !isSkippableCommand(name) {
+			*bin = name
```
[the remark this answers is not in the document]
Comment (clintecker) [reply validator/lint_tool_helpers.go:82]: Fixed in 7b21493. Removed `command` from `shellBuiltins`. Now handled specially in `callExprBinary`: `command -v foo` → "" (query), `command foo` → "foo" (execution), `command -p foo` → "foo" (default PATH execution).

File: validator/lint_tool_helpers.go [hunk 11]
```diff
@@ -77,13 +107,27 @@ var shellBuiltins = map[string]bool{
...
 // isShellBuiltin returns true if the command is a shell builtin.
 func isShellBuiltin(cmd string) bool {
 	return shellBuiltins[cmd]
 }
 
+// isSkippableCommand returns true if the command should be skipped
+// when looking for the primary tool binary (builtins + preamble).
+func isSkippableCommand(cmd string) bool {
+	return shellBuiltins[cmd] || preambleCommands[cmd]
+}
```
[the remark this answers is not in the document]
Comment (clintecker) [reply validator/lint_tool_helpers.go]: Fixed in 7b21493. Narrowed `preambleCommands` to just `mkdir` — matches DIP125 docs. `cp`, `mv`, `rm`, `touch`, `chmod` are no longer skipped.

File: validator/lint_tool_helpers.go [hunk 12]
```diff
@@ -77,13 +107,27 @@ var shellBuiltins = map[string]bool{
...
 	"trap": true, "wait": true, "true": true, "false": true,
 	"source": true, ".": true, "local": true, "declare": true,
+	"set": true, "cd": true, "export": true, "unset": true,
+	"command": true,
+}
+
+// preambleCommands are external binaries commonly used for setup that
+// should be skipped when finding the "real" tool binary in a command block.
+var preambleCommands = map[string]bool{
+	"mkdir": true, "touch": true, "chmod": true, "rm": true, "cp": true, "mv": true,
```
[the remark this answers is not in the document]
Comment (clintecker) [reply validator/lint_tool_helpers.go]: Fixed in 7b21493. Preamble narrowed to just `mkdir`, matching docs.

File: validator/lint_tool_helpers.go [hunk 13]
```diff
@@ -42,30 +43,59 @@ func checkToolCtxVars(n *ir.Node) []Diagnostic {
...
+	parser := syntax.NewParser(syntax.KeepComments(false))
+	prog, err := parser.Parse(strings.NewReader(command), "")
+	if err != nil {
+		return ""
+	}
+	var bin string
+	syntax.Walk(prog, walkForBinary(&bin))
+	return bin
+}
+
```
[the remark this answers is not in the document]
Comment (clintecker) [reply validator/lint_tool_helpers.go]: Fixed in 7b21493. Added `extractBinaryFallback` — on parse errors, falls back to first non-skippable token. Preserves DIP125 signal for commands the AST parser can't handle.
